README.md
# clue/reactphp-ssh-proxy

[![CI status](https://github.com/clue/reactphp-ssh-proxy/actions/workflows/ci.yml/badge.svg)](https://github.com/clue/reactphp-ssh-proxy/actions)
[![installs on Packagist](https://img.shields.io/packagist/dt/clue/reactphp-ssh-proxy?color=blue&label=installs%20on%20Packagist)](https://packagist.org/packages/clue/reactphp-ssh-proxy)

Async SSH proxy connector and forwarder, tunnel any TCP/IP-based protocol through an SSH server,
built on top of [ReactPHP](https://reactphp.org/).

[Secure Shell (SSH)](https://en.wikipedia.org/wiki/Secure_Shell) is a secure
network protocol that is most commonly used to access a login shell on a remote
server. Its architecture allows it to use multiple secure channels over a single
connection. Among others, this can also be used to create an "SSH tunnel", which
is commonly used to tunnel HTTP(S) traffic through an intermediary ("proxy"), to
conceal the origin address (anonymity) or to circumvent address blocking
(geoblocking). This can be used to tunnel any TCP/IP-based protocol (HTTP, SMTP,
IMAP etc.) and as such also allows you to access local services that are otherwise
not accessible from the outside (database behind firewall).
This library is implemented as a lightweight process wrapper around the `ssh` client
binary and provides a simple API to create these tunneled connections for you.
Because it implements ReactPHP's standard
[`ConnectorInterface`](https://github.com/reactphp/socket#connectorinterface),
it can simply be used in place of a normal connector.
This makes it fairly simple to add SSH proxy support to pretty much any
existing higher-level protocol implementation.

* **Async execution of connections** -
  Send any number of SSH proxy requests in parallel and process their
  responses as soon as results come in.
  The Promise-based design provides a *sane* interface to working with out of
  order responses and possible connection errors.
* **Standard interfaces** -
  Allows easy integration with existing higher-level components by implementing
  ReactPHP's standard
  [`ConnectorInterface`](https://github.com/reactphp/socket#connectorinterface).
* **Lightweight, SOLID design** -
  Provides a thin abstraction that is [*just good enough*](https://en.wikipedia.org/wiki/Principle_of_good_enough)
  and does not get in your way.
  Builds on top of well-tested components and well-established concepts instead of reinventing the wheel.
* **Good test coverage** -
  Comes with an automated tests suite and is regularly tested against actual SSH servers in the wild.

**Table of contents**

* [Support us](#support-us)
* [Quickstart example](#quickstart-example)
* [API](#api)
    * [SshProcessConnector](#sshprocessconnector)
    * [SshSocksConnector](#sshsocksconnector)
* [Usage](#usage)
    * [Plain TCP connections](#plain-tcp-connections)
    * [Secure TLS connections](#secure-tls-connections)
    * [HTTP requests](#http-requests)
    * [Database tunnel](#database-tunnel)
    * [Connection timeout](#connection-timeout)
    * [DNS resolution](#dns-resolution)
    * [Password authentication](#password-authentication)
* [Install](#install)
* [Tests](#tests)
* [License](#license)
* [More](#more)

## Support us

We invest a lot of time developing, maintaining and updating our awesome
open-source projects. You can help us sustain this high-quality of our work by
[becoming a sponsor on GitHub](https://github.com/sponsors/clue). Sponsors get
numerous benefits in return, see our [sponsoring page](https://github.com/sponsors/clue)
for details.

Let's take these projects to the next level together! 🚀

## Quickstart example

The following example code demonstrates how this library can be used to send a
secure HTTPS request to google.com through a remote SSH server:

```php
<?php

require __DIR__ . '/vendor/autoload.php';

$proxy = new Clue\React\SshProxy\SshProcessConnector('alice@example.com');

$connector = new React\Socket\Connector(array(
    'tcp' => $proxy,
    'dns' => false
));

$browser = new React\Http\Browser($connector);

$browser->get('https://google.com/')->then(function (Psr\Http\Message\ResponseInterface $response) {
    var_dump($response->getHeaders(), (string) $response->getBody());
}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;
});
```

See also the [examples](examples).

## API

### SshProcessConnector

The `SshProcessConnector` is responsible for creating plain TCP/IP connections to
any destination by using an intermediary SSH server as a proxy server.

```
[you] -> [proxy] -> [destination]
```

This class is implemented as a lightweight process wrapper around the `ssh`
client binary, so it will spawn one `ssh` process for each connection. For
example, if you [open a connection](#plain-tcp-connections) to
`tcp://reactphp.org:80`, it will run the equivalent of `ssh -W reactphp.org:80 alice@example.com`
and forward data from its standard I/O streams. For this to work, you'll have to
make sure that you have a suitable SSH client installed. On Debian/Ubuntu-based
systems, you may simply install it like this:

```bash
sudo apt install openssh-client
```

Its constructor simply accepts an SSH proxy server URL:

```php
$proxy = new Clue\React\SshProxy\SshProcessConnector('alice@example.com');
```

The proxy URL may or may not contain a scheme and port definition. The default
port will be `22` for SSH, but you may have to use a custom port depending on
your SSH server setup.

This class takes an optional `LoopInterface|null $loop` parameter that can be used to
pass the event loop instance to use for this object. You can use a `null` value
here in order to use the [default loop](https://github.com/reactphp/event-loop#loop).
This value SHOULD NOT be given unless you're sure you want to explicitly use a
given event loop instance.

Keep in mind that this class is implemented as a lightweight process wrapper
around the `ssh` client binary and that it will spawn one `ssh` process for each
connection. If you open more connections, it will spawn one `ssh` process for
each connection. Each process will take some time to create a new SSH connection
and then keep running until the connection is closed, so you're recommended to
limit the total number of concurrent connections. If you plan to only use a
single or few connections (such as a single database connection), using this
class is the recommended approach. If you plan to create multiple connections or
have a larger number of connections (such as an HTTP client), you're recommended
to use the [`SshSocksConnector`](#sshsocksconnector) instead.

This is one of the two main classes in this package.
Because it implements ReactPHP's standard
[`ConnectorInterface`](https://github.com/reactphp/socket#connectorinterface),
it can simply be used in place of a normal connector.
Accordingly, it provides only a single public method, the
[`connect()`](https://github.com/reactphp/socket#connect) method.
The `connect(string $uri): PromiseInterface<ConnectionInterface, Exception>`
method can be used to establish a streaming connection.
It returns a [Promise](https://github.com/reactphp/promise) which either
fulfills with a [ConnectionInterface](https://github.com/reactphp/socket#connectioninterface)
on success or rejects with an `Exception` on error.

This makes it fairly simple to add SSH proxy support to pretty much any
higher-level component:

```diff
- $acme = new AcmeApi($connector);
+ $proxy = new Clue\React\SshProxy\SshProcessConnector('alice@example.com');
+ $acme = new AcmeApi($proxy);
```

### SshSocksConnector

The `SshSocksConnector` is responsible for creating plain TCP/IP connections to
any destination by using an intermediary SSH server as a proxy server.

```
[you] -> [proxy] -> [destination]
```

This class is implemented as a lightweight process wrapper around the `ssh`
client binary and it will spawn one `ssh` process on demand for multiple
connections. For example, once you [open a connection](#plain-tcp-connections)
to `tcp://reactphp.org:80` for the first time, it will run the equivalent of
`ssh -D 1080 alice@example.com` to run the SSH client in local SOCKS proxy server
mode and will then create a SOCKS client connection to this server process. You
can create any number of connections over this one process and it will keep this
process running while there are any open connections and will automatically
close it when it is idle. For this to work, you'll have to make sure that you
have a suitable SSH client installed. On Debian/Ubuntu-based systems, you may
simply install it like this:

```bash
sudo apt install openssh-client
```

Its constructor simply accepts an SSH proxy server URL:

```php
$proxy = new Clue\React\SshProxy\SshSocksConnector('alice@example.com');
```

The proxy URL may or may not contain a scheme and port definition. The default
port will be `22` for SSH, but you may have to use a custom port depending on
your SSH server setup.

This class takes an optional `LoopInterface|null $loop` parameter that can be used to
pass the event loop instance to use for this object. You can use a `null` value
here in order to use the [default loop](https://github.com/reactphp/event-loop#loop).
This value SHOULD NOT be given unless you're sure you want to explicitly use a
given event loop instance.

Keep in mind that this class is implemented as a lightweight process wrapper
around the `ssh` client binary and that it will spawn one `ssh` process for
multiple connections. This process will take some time to create a new SSH
connection and then keep running until the last connection is closed. If you
plan to create multiple connections or have a larger number of concurrent
connections (such as an HTTP client), using this class is the recommended
approach. If you plan to only use a single or few connections (such as a single
database connection), you're recommended to use the [`SshProcessConnector`](#sshprocessconnector)
instead.

This class defaults to spawning the SSH client process in SOCKS proxy server
mode listening on `127.0.0.1:1080`. If this port is already in use or if you want
to use multiple instances of this class to connect to different SSH proxy
servers, you may optionally pass a unique bind address like this:

```php
$proxy = new Clue\React\SshProxy\SshSocksConnector('alice@example.com?bind=127.1.1.1:1081');
```

> *Security note for multi-user systems*: This class will spawn the SSH client
  process in local SOCKS server mode and will accept connections only on the
  localhost interface by default. If you're running on a multi-user system,
  other users on the same system may be able to connect to this proxy server and
  create connections over it. If this applies to your deployment, you're
  recommended to use the [`SshProcessConnector](#sshprocessconnector) instead
  or set up custom firewall rules to prevent unauthorized access to this port.

This is one of the two main classes in this package.
Because it implements ReactPHP's standard
[`ConnectorInterface`](https://github.com/reactphp/socket#connectorinterface),
it can simply be used in place of a normal connector.
Accordingly, it provides only a single public method, the
[`connect()`](https://github.com/reactphp/socket#connect) method.
The `connect(string $uri): PromiseInterface<ConnectionInterface, Exception>`
method can be used to establish a streaming connection.
It returns a [Promise](https://github.com/reactphp/promise) which either
fulfills with a [ConnectionInterface](https://github.com/reactphp/socket#connectioninterface)
on success or rejects with an `Exception` on error.

This makes it fairly simple to add SSH proxy support to pretty much any
higher-level component:

```diff
- $acme = new AcmeApi($connector);
+ $proxy = new Clue\React\SshProxy\SshSocksConnector('alice@example.com');
+ $acme = new AcmeApi($proxy);
```

## Usage

### Plain TCP connections

SSH proxy servers are commonly used to issue HTTPS requests to your destination.
However, this is actually performed on a higher protocol layer and this
project is actually inherently a general-purpose plain TCP/IP connector.
As documented above, you can simply invoke the `connect()` method to establish
a streaming plain TCP/IP connection on the `SshProcessConnector` or `SshSocksConnector`
and use any higher level protocol like so:

```php
$proxy = new Clue\React\SshProxy\SshProcessConnector('alice@example.com');
// or
$proxy = new Clue\React\SshProxy\SshSocksConnector('alice@example.com');

$proxy->connect('tcp://smtp.googlemail.com:587')->then(function (React\Socket\ConnectionInterface $connection) {
    $connection->write("EHLO local\r\n");
    $connection->on('data', function ($chunk) use ($connection) {
        echo $chunk;
    });
});
```

You can either use the `SshProcessConnector` or `SshSocksConnector` directly or you
may want to wrap this connector in ReactPHP's [`Connector`](https://github.com/reactphp/socket#connector):

```php
$proxy = new Clue\React\SshProxy\SshProcessConnector('alice@example.com');
// or
$proxy = new Clue\React\SshProxy\SshSocksConnector('alice@example.com');

$connector = new React\Socket\Connector(array(
    'tcp' => $proxy,
    'dns' => false
));

$connector->connect('tcp://smtp.googlemail.com:587')->then(function (React\Socket\ConnectionInterface $connection) {
    $connection->write("EHLO local\r\n");
    $connection->on('data', function ($chunk) use ($connection) {
        echo $chunk;
    });
});
```

For this example, you can use either the `SshProcessConnector` or `SshSocksConnector`.
Keep in mind that this project is implemented as a lightweight process wrapper
around the `ssh` client binary. While the `SshProcessConnector` will spawn one
`ssh` process for each connection, the `SshSocksConnector` will spawn one `ssh`
process that will be shared for multiple connections, see also above for more
details.

### Secure TLS connections

The `SshSocksConnector` can also be used if you want to establish a secure TLS connection
(formerly known as SSL) between you and your destination, such as when using
secure HTTPS to your destination site. You can simply wrap this connector in
ReactPHP's [`Connector`](https://github.com/reactphp/socket#connector):

```php
$proxy = new Clue\React\SshProxy\SshSocksConnector('alice@example.com');

$connector = new React\Socket\Connector(array(
    'tcp' => $proxy,
    'dns' => false
));

$connector->connect('tls://smtp.googlemail.com:465')->then(function (React\Socket\ConnectionInterface $connection) {
    $connection->write("EHLO local\r\n");
    $connection->on('data', function ($chunk) use ($connection) {
        echo $chunk;
    });
});
```

> Note how secure TLS connections are in fact entirely handled outside of
  this SSH proxy client implementation.
  The `SshProcessConnector` does not currently support secure TLS connections
  because PHP's underlying crypto functions require a socket resource and do not
  work for virtual connections. As an alternative, you're recommended to use the
  `SshSocksConnector` as given in the above example.

### HTTP requests

This library also allows you to send
[HTTP requests through an SSH proxy server](https://github.com/reactphp/http#ssh-proxy).

In order to send HTTP requests, you first have to add a dependency for
[ReactPHP's async HTTP client](https://github.com/reactphp/http#client-usage).
This allows you to send both plain HTTP and TLS-encrypted HTTPS requests like this:

```php
$proxy = new Clue\React\SshProxy\SshSocksConnector('alice@example.com');

$connector = new React\Socket\Connector(array(
    'tcp' => $proxy,
    'dns' => false
));

$browser = new React\Http\Browser($connector);

$browser->get('https://google.com/')->then(function (Psr\Http\Message\ResponseInterface $response) {
    var_dump($response->getHeaders(), (string) $response->getBody());
}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;
});
```

We recommend using the `SshSocksConnector`, this works for both plain HTTP
and TLS-encrypted HTTPS requests. When using the `SshProcessConnector`, this only
works for plaintext HTTP requests.

See also [ReactPHP's HTTP client](https://github.com/reactphp/http#client-usage)
and any of the [examples](examples) for more details.

### Database tunnel

We should now have a basic understanding of how we can tunnel any TCP/IP-based
protocol over an SSH proxy server. Besides using this to access "external"
services, this is also particularly useful because it allows you to access
network services otherwise only local to this SSH server from the outside, such
as a firewalled database server.

For example, this allows us to combine an
[async MySQL database client](https://github.com/friends-of-reactphp/mysql) and
the above SSH proxy server setup, so we can access a firewalled MySQL database
server through an SSH tunnel. Here's the gist:

```php
$proxy = new Clue\React\SshProxy\SshProcessConnector('alice@example.com');

$uri = 'test:test@localhost/test';
$factory = new React\MySQL\Factory(null, $proxy);
$connection = $factory->createLazyConnection($uri);

$connection->query('SELECT * FROM book')->then(
    function (React\MySQL\QueryResult $command) {
        echo count($command->resultRows) . ' row(s) in set' . PHP_EOL;
    },
    function (Exception $error) {
        echo 'Error: ' . $error->getMessage() . PHP_EOL;
    }
);

$connection->quit();
```

See also [example #21](examples) for more details.

This example will automatically launch the `ssh` client binary to create the
connection to a database server that can not otherwise be accessed from the
outside. From the perspective of the database server, this looks just like a
regular, local connection. From this code's perspective, this will create a
regular, local connection which just happens to use a secure SSH tunnel to
transport this to a remote server, so you can send any query like you would to a
local database server.

### Connection timeout

By default, neither the `SshProcessConnector` nor the `SshSocksConnector` implement
any timeouts for establishing remote connections.
Your underlying operating system may impose limits on pending and/or idle TCP/IP
connections, anywhere in a range of a few minutes to several hours.

Many use cases require more control over the timeout and likely values much
smaller, usually in the range of a few seconds only.

You can use ReactPHP's [`Connector`](https://github.com/reactphp/socket#connector)
to decorate any given `ConnectorInterface` instance.
It provides the same `connect()` method, but will automatically reject the
underlying connection attempt if it takes too long:

```php
$proxy = new Clue\React\SshProxy\SshProcessConnector('alice@example.com');
// or
$proxy = new Clue\React\SshProxy\SshSocksConnector('alice@example.com');

$connector = new React\Socket\Connector(array(
    'tcp' => $proxy,
    'dns' => false,
    'timeout' => 3.0
));

$connector->connect('tcp://google.com:80')->then(function (React\Socket\ConnectionInterface $connection) {
    // connection succeeded within 3.0 seconds
});
```

See also any of the [examples](examples).

> Note how the connection timeout is in fact entirely handled outside of this
  SSH proxy client implementation.

### DNS resolution

By default, neither the `SshProcessConnector` nor the `SshSocksConnector` perform
any DNS resolution at all and simply forward any hostname you're trying to
connect to the remote proxy server. The remote proxy server is thus responsible
for looking up any hostnames via DNS (this default mode is thus called *remote DNS resolution*).

As an alternative, you can also send the destination IP to the remote proxy
server.
In this mode you either have to stick to using IPs only (which is ofen unfeasable)
or perform any DNS lookups locally and only transmit the resolved destination IPs
(this mode is thus called *local DNS resolution*).

The default *remote DNS resolution* is useful if your local `SshProcessConnector` 
or `SshSocksConnector` either can not resolve target hostnames because it has no
direct access to the internet or if it should not resolve target hostnames
because its outgoing DNS traffic might be intercepted.

As noted above, the `SshProcessConnector` and `SshSocksConnector` default to using
remote DNS resolution. However, wrapping them in ReactPHP's
[`Connector`](https://github.com/reactphp/socket#connector) actually
performs local DNS resolution unless explicitly defined otherwise.
Given that remote DNS resolution is assumed to be the preferred mode, all
other examples explicitly disable DNS resolution like this:

```php
$proxy = new Clue\React\SshProxy\SshProcessConnector('alice@example.com');
// or
$proxy = new Clue\React\SshProxy\SshSocksConnector('alice@example.com');

$connector = new React\Socket\Connector(array(
    'tcp' => $proxy,
    'dns' => false
));
```

If you want to explicitly use *local DNS resolution*, you can use the following code:

```php
$proxy = new Clue\React\SshProxy\SshProcessConnector('alice@example.com');
// or
$proxy = new Clue\React\SshProxy\SshSocksConnector('alice@example.com');

// set up Connector which uses Google's public DNS (8.8.8.8)
$connector = new React\Socket\Connector(array(
    'tcp' => $proxy,
    'dns' => '8.8.8.8'
));
```

> Note how local DNS resolution is in fact entirely handled outside of this
  SSH proxy client implementation.

### Password authentication

Note that this class is implemented as a lightweight process wrapper around the
`ssh` client binary. It works under the assumption that you have verified you
can access your SSH proxy server on the command line like this:

```bash
# test SSH access
ssh alice@example.com echo hello
```

Because this class is designed to be used to create any number of connections,
it does not provide a way to interactively ask for your password. Similarly,
the `ssh` client binary does not provide a way to "pass" in the password on the
command line for security reasons. This means that you are highly recommended to
set up pubkey-based authentication without a password for this to work best.

Additionally, this library provides a way to pass in a password in a somewhat
less secure way if your use case absolutely requires this. Before proceeding,
please consult your SSH documentation to find out why this may be a bad idea and
why pubkey-based authentication is usually the better alternative.
If your SSH proxy server requires password authentication, you may pass the
username and password as part of the SSH proxy server URL like this:

```php
$proxy = new Clue\React\SshProxy\SshProcessConnector('alice:password@example.com');
// or
$proxy = new Clue\React\SshProxy\SshSocksConnector('alice:password@example.com');
```

For this to work, you will have to have the `sshpass` binary installed. On
Debian/Ubuntu-based systems, you may simply install it like this:

```bash
sudo apt install sshpass
```

Note that both the username and password must be percent-encoded if they contain
special characters:

```php
$user = 'he:llo';
$pass = 'p@ss';
$url = rawurlencode($user) . ':' . rawurlencode($pass) . '@example.com';

$proxy = new Clue\React\SshProxy\SshProcessConnector($url);
```

## Install

The recommended way to install this library is [through Composer](https://getcomposer.org/).
[New to Composer?](https://getcomposer.org/doc/00-intro.md)

This project follows [SemVer](https://semver.org/).
This will install the latest supported version:

```bash
composer require clue/reactphp-ssh-proxy:^1.4
```

See also the [CHANGELOG](CHANGELOG.md) for details about version upgrades.

This project aims to run on any platform and thus does not require any PHP
extensions and supports running on legacy PHP 5.3 through current PHP 8+ and
HHVM.
It's *highly recommended to use the latest supported PHP version* for this project.

This project is implemented as a lightweight process wrapper around the `ssh`
client binary, so you'll have to make sure that you have a suitable SSH client
installed. On Debian/Ubuntu-based systems, you may simply install it like this:

```bash
sudo apt install openssh-client
```

Additionally, if you use [password authentication](#password-authentication)
(not recommended), then you will have to have the `sshpass` binary installed. On
Debian/Ubuntu-based systems, you may simply install it like this:

```bash
sudo apt install sshpass
```

*Running on [Windows is currently not supported](https://github.com/clue/reactphp-ssh-proxy/issues/12)*

## Tests


CHANGELOG.md
# Changelog

## 1.4.0 (2022-09-02)

*   Feature: Full support for PHP 8.1 and PHP 8.2.
    (#32 by @clue and #37 by @SimonFrings)

*   Feature: Mark passwords and URIs as `#[\SensitiveParameter]` (PHP 8.2+).
    (#38 by @SimonFrings)

*   Feature: Forward compatibility with upcoming Promise v3.
    (#35 by @clue)

*   Feature: Avoid dependency on `ext-filter`.
    (#31 by @clue)

*   Improve test suite and use new reactphp/async package instead of clue/reactphp-block.
    (#36 and #39 by @SimonFrings)

## 1.3.0 (2021-08-06)

*   Feature: Simplify usage by supporting new default loop.
    (#27 and #28 by @clue)

    ```php
    // old (still supported)
    $proxy = new Clue\React\SshProxy\SshProcessConnector('user@example.com', $loop);
    $proxy = new Clue\React\SshProxy\SshSocksConnector('user@example.com', $loop);

    // new (using default loop)
    $proxy = new Clue\React\SshProxy\SshProcessConnector('user@example.com');
    $proxy = new Clue\React\SshProxy\SshSocksConnector('user@example.com');
    ```

*   Documentation improvements and updated examples.
    (#25, #29 and #30 by @clue and #23 and #26 by @SimonFrings)

*   Improve test suite and use GitHub actions for continuous integration (CI).
    (#24 by @SimonFrings)

## 1.2.0 (2020-10-23)

*   Fix: Fix error reporting when parsing invalid SSH server URL.
    (#15 by @clue)

*   Enhanced documentation for ReactPHP's new HTTP client.
    (#21 by @SimonFrings)

*   Improve test suite and add `.gitattributes` to exclude dev files from exports.
    Prepare PHP 8 support, update to PHPUnit 9 and simplify test matrix.
    (#14 by @clue and #16, #18, #19 and 22 by @SimonFrings)

## 1.1.1 (2019-05-01)

*   Fix: Only start consuming STDOUT data once connection is ready.
    (#11 by @clue)

*   Add documentation and example for MySQL database SSH tunnel.
    (#13 by @clue)

## 1.1.0 (2019-01-21)

*   Feature: Improve platform support (chroot environments, Mac and others) and
    do not inherit open FDs to SSH child process by overwriting and closing.
    (#10 by @clue)

## 1.0.0 (2018-12-19)

*   First stable release, following SemVer


src/SshSocksConnector.php
<?php

namespace Clue\React\SshProxy;

use Clue\React\Socks\Client;
use Clue\React\SshProxy\Io\LineSeparatedReader;
use React\ChildProcess\Process;
use React\EventLoop\Loop;
use React\EventLoop\LoopInterface;
use React\Promise\Deferred;
use React\Socket\ConnectorInterface;
use React\Socket\TcpConnector;
use React\Socket\ConnectionInterface;

class SshSocksConnector implements ConnectorInterface
{
    private $cmd;

    /** @var LoopInterface */
    private $loop;

    private $debug = false;

    private $bind = '127.0.0.1:1080';
    private $socks;

    private $listen;
    private $pending = 0;
    private $idlePeriod = 0.001;
    private $idleTimer;

    /**
     *
     * [ssh://][user[:pass]@]host[:port]
     *
     * You're highly recommended to use public keys instead of passing a
     * password here. If you really need to pass a password, please be aware
     * that this will be passed as a command line argument to `sshpass`
     * (which may need to be installed) and this password may leak to the
     * process list if other users have access to your system.
     *
     * This class takes an optional `LoopInterface|null $loop` parameter that can be used to
     * pass the event loop instance to use for this object. You can use a `null` value
     * here in order to use the [default loop](https://github.com/reactphp/event-loop#loop).
     * This value SHOULD NOT be given unless you're sure you want to explicitly use a
     * given event loop instance.
     *
     * @param string $uri
     * @param ?LoopInterface $loop
     * @throws \InvalidArgumentException
     */
    public function __construct($uri, LoopInterface $loop = null)
    {
        // URI must use optional ssh:// scheme, must contain host and neither pass nor target must start with dash
        $parts = \parse_url((\strpos($uri, '://') === false ? 'ssh://' : '') . $uri);
        $pass = isset($parts['pass']) ? \rawurldecode($parts['pass']) : null;
        $target = (isset($parts['user']) ? \rawurldecode($parts['user']) . '@' : '') . (isset($parts['host']) ? $parts['host'] : '');
        if (!isset($parts['scheme'], $parts['host']) || $parts['scheme'] !== 'ssh' || (isset($pass[0]) && $pass[0] === '-') || $target[0] === '-') {
            throw new \InvalidArgumentException('Invalid SSH server URI');
        }

        $this->cmd = 'exec ';
        if ($pass !== null) {
            $this->cmd .= 'sshpass -p ' . \escapeshellarg($pass) . ' ';
        }
        $this->cmd .= 'ssh -v -o ExitOnForwardFailure=yes -N ';

        // disable interactive password prompt if no password was given (see sshpass above)
        if ($pass === null) {
            $this->cmd .= '-o BatchMode=yes ';
        }

        if (isset($parts['port']) && $parts['port'] !== 22) {
            $this->cmd .= '-p ' . $parts['port'] . ' ';
        }

        $args = array();
        \parse_str((string) parse_url($uri, \PHP_URL_QUERY), $args);
        if (isset($args['bind'])) {
            $parts = parse_url('tcp://' . $args['bind']);
            if (!isset($parts['scheme'], $parts['host'], $parts['port']) || @\inet_pton(\trim($parts['host'], '[]')) === false) {
                throw new \InvalidArgumentException('Invalid bind address given');
            }
            $this->bind = $args['bind'];
        }

        $this->loop = $loop ?: Loop::get();
        $this->socks = new Client('socks4://' . $this->bind, new TcpConnector($this->loop));
        $this->cmd .= '-D ' . \escapeshellarg($this->bind) . ' ' . \escapeshellarg($target);
    }

    public function connect($uri)
    {
        // URI must use optional tcp:// scheme, must contain host and port
        $parts = \parse_url((\strpos($uri, '://') === false ? 'tcp://' : '') . $uri);
        if (!isset($parts['scheme'], $parts['host'], $parts['port']) || $parts['scheme'] !== 'tcp') {
            return \React\Promise\reject(new \InvalidArgumentException('Invalid target URI'));
        }

        // no listening SSH client process started? start one and remember reference
        // if spawning SSH client process fails, clear reference for next start
        if ($this->listen === null) {
            $listen =& $this->listen;
            $this->listen = $this->listen()->then(null, function ($e) use (&$listen) {
                $listen = null;
                throw $e;
            });

            // @codeCoverageIgnoreStart
            if ($this->debug) {
                echo 'Starting to listen' . PHP_EOL;
                $this->listen->then(function () {
                    echo 'Listening' . PHP_EOL;
                }, function () {
                    echo 'Failed listening' . PHP_EOL;
                });
            }
            // @codeCoverageIgnoreEnd
        }

        // keep track of pending connection attempts to close process on idle again
        $this->awake();
        $that = $this;
        $socks = $this->socks;

        $connecting = null;
        $cancelled = false;
        $deferred = new Deferred(function () use ($uri, $that, &$connecting, &$cancelled) {
            // cancel pending SOCKS connection attempt if SSH client process is ready
            if ($connecting !== null) {
                $connecting->cancel();
                $connecting = null;
                return;
            }

            // otherwise just reject and keep starting SSH client process until idle timeout
            $cancelled = true;
            $that->idle();
            throw new \RuntimeException('Connection to ' . $uri . ' cancelled while waiting for SSH client');
        });

        $this->listen->then(function () use ($uri, $that, $socks, $deferred, &$connecting) {
            // SSH client process started => start SOCKS connection
            $connecting = $socks->connect($uri)->then(function (ConnectionInterface $connection) use ($that, $deferred) {
                // connection to SOCKS server successful => consider idle when connection closes again
                $connection->on('close', function () use ($that) {
                    $that->idle();
                });
                $deferred->resolve($connection);
            }, function (\Exception $e) use ($that, $deferred) {
                // connection to SOCKS server failed => reject
                $that->idle();

                $deferred->reject($e);
            });
        }, function (\Exception $e) use ($that, $uri, $deferred, &$cancelled) {
            // creating SOCKS proxy failed => reject (unless connection attempt has already been cancelled)
            if ($cancelled) {
                return;
            }

            $that->idle();

            $deferred->reject(new \RuntimeException(
                'Connection to ' . $uri . ' failed because SSH client process died (' . $e->getMessage() . ')'
            ));
        });

        return $deferred->promise();
    }

    /** @internal */
    public function awake()
    {
        ++$this->pending;

        if ($this->idleTimer !== null) {
            $this->loop->cancelTimer($this->idleTimer);
            $this->idleTimer = null;
        }
    }

    /** @internal */
    public function idle()
    {
        --$this->pending;

        if ($this->pending < 1 && $this->idleTimer === null && $this->listen !== null) {
            $listen =& $this->listen;
            $debug = $this->debug;
            $this->idleTimer = $this->loop->addTimer($this->idlePeriod, function () use (&$listen, $debug) {
                $listen->then(function (Process $process) {
                    $process->terminate();
                });
                $listen->cancel();
                $listen = null;

                if ($debug) {
                    echo 'Stopped listening' . PHP_EOL; // @codeCoverageIgnore
                }
            });
        }
    }

    private function listen()
    {
        $process = Io\processWithoutFds($this->cmd);
        $process->start($this->loop);

        $deferred = new Deferred(function () use ($process) {
            $process->stdin->close();
            $process->terminate();
        });

        // process STDERR one line at a time
        $last = 'n/a';
        $debug = $this->debug;
        $stderr = new LineSeparatedReader($process->stderr);
        $stderr->on('data', function ($line) use ($deferred, $process, &$last, $debug) {
            // remember last line for error output in case process exits
            // forwarding error will be printed right to stderr and then exit (see below)
            // bind: Address already in use
            // Could not request local forwarding.
            $last = $line;

            if ($debug) {
                echo \addcslashes($line, "\0..\032") . PHP_EOL; // @codeCoverageIgnore
            }

            // match everything related to our forwarding channel
            // forwarding success will only be reported to stderr via debug1 / -v
            // debug1: channel 0: new [port listener]
            if (\preg_match('/^(?:debug\d\: )channel \d: new \[port listener\]$/', $line)) {
                $deferred->resolve($process);
            }
        });

        $process->on('exit', function ($code) use ($deferred, &$last) {
            $deferred->reject(new \RuntimeException(
                $last,
                $code
            ));
        });

        return $deferred->promise();
    }
}
